endpoints/put.go

```diff
@@ -142,6 +106,105 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
 	}
 }
 
+func validateAndEncode(backend *backendCallObject, exit *exitInfo) {
+	for index, p := range backend.put.Puts {
+		var toCache string
+
+		if len(p.Value) == 0 {
+			exit.errMsg = "Missing value."
+			exit.status = http.StatusBadRequest
+			return
+		}
+		if p.TTLSeconds < 0 {
+			exit.errMsg = fmt.Sprintf("request.puts[%d].ttlseconds must not be negative.", p.TTLSeconds)
+			exit.status = http.StatusBadRequest
+			return
+		}
+
+		if p.Type == backends.XML_PREFIX {
+			if p.Value[0] != byte('"') || p.Value[len(p.Value)-1] != byte('"') {
+				exit.errMsg = fmt.Sprintf("XML messages must have a String value. Found %v", p.Value)
+				exit.status = http.StatusBadRequest
+				return
+			}
+
+			// Be careful about the the cross-script escaping issues here. JSON requires quotation marks to be escaped,
+			// for example... so we'll need to un-escape it before we consider it to be XML content.
+			var interpreted string
+			json.Unmarshal(p.Value, &interpreted)
+			toCache = p.Type + interpreted
+		} else if p.Type == backends.JSON_PREFIX {
+			toCache = p.Type + string(p.Value)
+		} else {
+			exit.errMsg = fmt.Sprintf("Type must be one of [\"json\", \"xml\"]. Found %v", p.Type)
+			exit.status = http.StatusBadRequest
+			return
+		}
+		logrus.Debugf("Storing value: %s", toCache)
+		u2, err := uuid.NewV4()
+		if err != nil {
+			exit.errMsg = "Error generating version 4 UUID"
+			exit.status = http.StatusInternalServerError
+			return
+		}
+		backend.resp.Responses[index].UUID = u2.String()
+		backend.resp.toCacheStrings[index] = toCache
+	}
+	return
+}
+
+type backendCallObject struct {
+	client    backends.Backend
+	put       *PutRequest
+	resp      *PutResponse
+	allowKeys bool
+	ctx       context.Context
+}
+
+func callBackendInParallel(backend *backendCallObject, exit *exitInfo, index int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	callBackendGet(backend, index)
+	callBackendPut(backend, exit, index)
+}
+
+func callBackendGet(backend *backendCallObject, index int) {
+	if backend.allowKeys && len(backend.put.Puts[index].Key) > 0 {
+		s, err := backend.client.Get(backend.ctx, backend.put.Puts[index].Key)
+		if err != nil || len(s) == 0 {
+			backend.resp.Responses[index].UUID = backend.put.Puts[index].Key
+		} else {
+			backend.resp.Responses[index].UUID = ""
+		}
+	}
+}
+
+func callBackendPut(backend *backendCallObject, exit *exitInfo, index int) {
+	if len(backend.resp.Responses[index].UUID) > 0 {
```

Comment:
What's happening here is a bit hard to follow. I think you refactored the code well and this was hard to follow before as well. I think we can leave it alone for now, but consider revisiting in a future story to make it clear that this handles the case of a Key was specified but does not exist. Should be handled as an err instead of empty string.

Reply:
Totally agree. I tried to leave the logic as is and only focus in the parallelism part, but I agree that the `if err != nil || len(s) == 0` statment could be dissected to differentiate between one option or the other. Maybe, in case of an error, throw it and interrupt execution if the data service is down.